### DISCUSSION TITLE
Error in using custom eslint plugin rules on publishing it as npm package

### DISCUSSION THREAD

**Discussion by urvish-pujara** [eslint/eslint#18072] [Q&A] [answered]
![code](https://github.com/eslint/eslint/assets/83078848/55dc4c27-51ef-4d7e-a4ce-97a7dbda79ba)
I want to publish my eslint custom plugin as an npm package and use it in my ts based web-app.
So for the initial test i did the following:
- Made a small rule to avoid a variable and published it as npm package 
- Installed that package in my web-app
- Added the rule in .eslintrc.js as follows:
```
module.exports = {
    parser: '@typescript-eslint/parser',
    plugins: ['test-rule-1925'],
    rules: {
        'no-remedo-variable': "error",
    },
};
```
But running lint says ```Definition for rule 'no-remedo-variable' was not found```
To test the npm package i used use ```npm install eslint-plugin-test-rule-1925```

eslint version: 8.56.0
node version: 18.19.0
npm version: 10.2.3

I am trying to fix this error since a very long time. Can someone please look into it and tell me what am i missing here.

**Comment by Rec0iL99** [comment]
Hi @urvish-pujara, there are a few problems with how you have implemented your plugin. I'd recommend following the [`Custom Rule Tutorial`](https://eslint.org/docs/latest/extend/custom-rule-tutorial) from the docs. This will help you build a plugin for ESLint for the first time. Feel free to come back here if you have more questions.

**Reply by urvish-pujara** [reply]
I followed https://yehezgun.com/articles/publish-eslint-rules-as-npm-package almost exactly still got the error. As the code is very short can you please specify on where exactly the error is? It would be great help.

**Reply by Rec0iL99** [reply, marked as answer]
The `index.js` is the main entry point for your plugin. This is where you would need to export the custom rule you wrote.

```js
// eslint-plugin-test-rule-1925/index.js
const rule = require("./custom-rules/no-remedo-variable");
module.exports = rule;
```

Now you can use your plugin in a project with ESLint configured. You would need to prefix any rule from a plugin with the plugin name so that ESLint knows where to look for that specific rule.

```js
// test-plugin/.eslintrc.js
module.exports = {
    parser: '@typescript-eslint/parser',
    plugins: ['test-rule-1925'],
    rules: {
        'test-rule-1925/no-remedo-variable': "error",
    },
};
```

This should now fix everything. I'd again recommend going through the [custom rule tutorial](https://eslint.org/docs/latest/extend/custom-rule-tutorial) and [create a plugin](https://eslint.org/docs/latest/extend/plugins) doc to get a better understanding of everything.
